src/Data/Replicator/Linear/Internal.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# OPTIONS_HADDOCK hide #-}
+
+module Data.Replicator.Linear.Internal (Replicator (..), consume, map, pure, (<*>), next, next#, extract, Elim (..)) where
+
+import Data.Arity.Linear.Internal.Arity
+import Data.Kind (Constraint, Type)
+import Data.Replicator.Linear.Internal.ReplicationStream (ReplicationStream (..))
+import qualified Data.Replicator.Linear.Internal.ReplicationStream as ReplicationStream
+import GHC.TypeLits
+import Prelude.Linear.Internal
+
+-- | @Replicator@ is used to linearly duplicate values in an efficient fashion.
+-- @Movable@ types are deep-copied only once (no matter how many copies are
+-- required), whereas non-@Movable@ values are duplicated using the internal
+-- construct @ReplicationStream@. @Replicator@ also have a linear
+-- @Data.Applicative@ instance.
```

Comment:
See https://stackoverflow.com/a/23858109
(`consume`, `dup2`) seems to be the definition of a comonoid.

Reply:
But I'm not sure that `Replicator` is a co-monoid though. But any `Dupable` type seems to be an instance of comonoid.

Reply:
To be completely exact, comonoids can be presented with `consume` and `dup2`. But it's not the only presentation. The `Dupable` class, is quite concretely the class of comonoids for the linear functions (and the usual product), it's a different presentation (if more redundant).

In particular `Replicator a` is a comonoid, since it implements `Dupable` (which means the same thing).

The presentation of comonoids in terms of `dupR` corresponds to a presentation of a coalgebraic structures as co-algebras of comonads (here, the comonad is `Replicator`). That being said, with a bit more thought into it, we've basically just defined `Replicator a` as `∃ x. Dupable x ∧ x %1 -> a`, which pretty much reads as “I'm the free `Dupable`”. So I guess that it's fairly unsurprising after all. And not very informative.